Isolation daemon's workers metrics

karitra opened the pull request:
Retrieve workers metrics from _Cocaine isolation daemon_ (aka [stout](https://github.com/noxiouz/stout)), register resulting signals within metrics service hub, show (some, probably after post-processing) in cocaine-tool info. Special attention should be paid to short living workers, their post-mortem metrics should be gathered and aggregated similarly.

#### TODO for first implementation:
 - [x] make metrics retrieving configurable
 - [x] correct per app aggregation for monotonically increasing values
 - [x] add protocol tests to isolation mock service
 - [x] signal in some form about workers metrics not updated for predefined period of time
 - [x] protocol encapsulated error messages processing 

#### Planned TODO (probably for separate pull request):
 - [ ] adjust and stabilize Cocaine <=> Isolation Daemon protocol: metrics names, types
 - [ ] make poll calls work in case of Isolation Daemon absence/disability (working, not tested)
 - [ ] check for possibility of external_t::metrics() concurrent access


Note that in current implementation isolation metrics polling is switched _off_ by default, to enable it set following parameters in `node` section of CocainRT configuration file:
```json
{
"services" : {
    "node" : {
       "args" : {
          "isolate_metrics:" : true,
          "isolate_metrics_poll_period_s" : 10
        }
    }
}
```

3Hren committed:
wip: initial release

3Hren commented on docker/isolate.hpp:
```diff
@@ -42,7 +42,7 @@ class docker_t:
     virtual
     std::unique_ptr<api::cancellation_t>
-    spool(std::shared_ptr<api::spool_handle_base_t> handler);
+    spool(std::shared_ptr<api::spool_handle_base_t> handle);
```
Leave all renaming stuff for a separate PR.

3Hren commented on node/include/cocaine/api/isolate.hpp:
```diff
@@ -91,6 +91,19 @@ struct spawn_handle_base_t {
 
+struct metrics_handle_base_t {
+    virtual
+    ~metrics_handle_base_t() {}
```
`= default`.

3Hren commented on node/src/isolate/external.cpp:
```diff
@@ -91,6 +91,24 @@ struct spawn_dispatch_t :
+
+    metrics_dispatch_t(const std::string &name, std::shared_ptr<api::metrics_handle_base_t> handle)
+        : dispatch(name),
+          metrics_handle(std::move(handle))
```
Seems like there is no need to store handle inside this struct.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -55,6 +83,185 @@ struct collector_t {
+
+        auto alive_uuids = pool.apply([&] (const pool_type& pool) {
+            std::vector<std::string> alive;
+            alive.resize(pool.size());
```
Reserve? Instead your alive pool is doubled.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -55,6 +83,185 @@ struct collector_t {
+};
+
+//
+// TODO: should that stuff be moved to separate compilation unit?
```
Yep. Engine is too big enough.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -55,6 +83,185 @@ struct collector_t {
+    const std::unique_ptr<cocaine::logging::logger_t> log;
+
+    std::set<std::string> purgatory;
+    mutable std::mutex prg_lock;
```
Didn't get the idea about unsynchronized access to the purgatory.

3Hren commented on node/include/cocaine/detail/service/node/engine.hpp:
```diff
@@ -74,6 +74,9 @@ class engine_t : public std::enable_shared_from_this<engine_t> {
     /// Statistics.
     stats_t stats;
 
+    class metrics_retriever_t;
```
There is no need to introduce impl in `detail` files.

karitra committed:
wip: refactoring garbage to separate pile; post-review fixes

karitra committed:
wip: pool observers list; metrics_retriever_t make and init sequence; add isometrics.cpp to build list

karitra committed:
wip: make engine_t::pool_observers container synchronized; isolation response processor

karitra committed:
wip: app-wide metrics aggregation; purge dead/removed workers stats

karitra committed:
wip: refit isolate_mock with metrics handle; metrics aggregation refactoring

karitra committed:
wip: coxx-wide config + metrics aggregation refactoring + error handling

 - coxx-wide config for metrics_retriever_t
 - almost correct app-wide aggregates for accumulated values (should be tested)
 - isolate daemon protocol response elementary error handling

karitra commented:
Protocol neither tested nor complete, overall implementation is still _moving target_, error responses must be processed, local integration tests with _stout_ are scheduled to be done, overall tests are in wish-list, but updates may land into another PR, so @3Hren, @antmat  PTAL.

3Hren commented on docker/isolate.cpp:
```diff
@@ -285,3 +285,8 @@ docker_t::spawn(const std::string& path,
 }
+
+void
+docker_t::metrics(const dynamic_t&, std::shared_ptr<api::metrics_handle_base_t>) const
```
I'd rather call `on_data({})` instead of ignoring.

3Hren commented on docker/isolate.hpp:
```diff
@@ -51,6 +51,10 @@ class docker_t:
           const api::env_t& environment,
           std::shared_ptr<api::spawn_handle_base_t> handle);
 
+    virtual
```
Either `virtual` or `override`.

3Hren commented on node/include/cocaine/idl/isolate.hpp:
```diff
@@ -80,6 +81,23 @@ struct isolate {
+        }
+
+        typedef boost::mpl::list<
+            dynamic_t
```
Comments.

3Hren commented on node/include/cocaine/idl/isolate.hpp:
```diff
@@ -80,6 +81,23 @@ struct isolate {
+        typedef isolate_tag dispatch_type;
+
+        typedef option_of<
+            dynamic_t
```
Comments.

3Hren commented on node/include/cocaine/idl/isolate.hpp:
```diff
@@ -106,13 +124,15 @@ struct isolate_spooled {
 
 template<>
 struct protocol<isolate_tag> {
+
```
Return back.

3Hren commented on node/src/isolate/external.cpp:
```diff
@@ -434,4 +485,27 @@ external_t::spawn(const std::string& path,
+
+    auto load = std::make_shared<metrics_load_t>(inner, query, handle);
+    // implicitly copy this->inner to a variable to capture it to avoid expiration of inner
+    auto _inner = inner;
```
It's already captured by `load`, isn't it?

3Hren commented on node/src/isolate/process.cpp:
```diff
@@ -431,5 +431,11 @@ process_t::spawn(const std::string& path,
+void
+process_t::metrics(const dynamic_t&, std::shared_ptr<api::metrics_handle_base_t>) const
+{
+    COCAINE_LOG_WARNING(m_log, "legacy_process::metrics not implemented");
```
Call handle instead.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -54,9 +61,30 @@ engine_t::engine_t(context_t& context,
 {
+    observers->emplace_back(observer);
+
+    if (loop) {
```
Why?

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -54,9 +61,30 @@ engine_t::engine_t(context_t& context,
+                profile.isolate.type,
+                profile.isolate.args);
+
+        if (isolate) {
```
We get either valid pointer or exception. This is unnecessary check.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -54,9 +61,30 @@ engine_t::engine_t(context_t& context,
-    observer(observer),
     stats(context, manifest_.name, std::chrono::seconds(2))
 {
+    observers->emplace_back(observer);
```
You have `attach_pool_observer`, use it!

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -296,6 +329,9 @@ auto engine_t::assign(slave_t& slave, load_t& load) -> void {
 
 auto engine_t::despawn(const std::string& id, despawn_policy_t policy) -> void {
+
+    auto was_despawned = false;
```
Return from lambda instead.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -305,14 +341,22 @@ auto engine_t::despawn(const std::string& id, despawn_policy_t policy) -> void {
+
+    if (was_despawned) {
+        observers.apply([&](const observers_type& observers) {
+            boost::for_each(observers, [&](const std::shared_ptr<pool_observer>& o) {
```
There is no profit for `boost::for_each` comparing with range-based for.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -345,11 +389,16 @@ auto engine_t::on_handshake(const std::string& id, std::shared_ptr<session_t> se
         });
+
+        observers.apply([&](const observers_type& observers) {
+            boost::for_each(observers, [](const std::shared_ptr<pool_observer>& o) {
```
The same here.

3Hren commented on node/src/node/engine.cpp:
```diff
@@ -394,7 +443,12 @@ auto engine_t::on_slave_death(const std::error_code& ec, std::string uuid) -> vo
 
-    observer.despawned();
+    observers.apply([&](const observers_type& observers) {
+        boost::for_each(observers, [&](const std::shared_ptr<pool_observer>& o) {
```
The same.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+    const std::vector<std::pair<std::string, bool>> counter_metrics_names =
+    {
+        // yet abstract cpu load measurement
+        { "cpu", false},
```
Use enums instead of bool. It improves readability. Thus, there will be no need for separate comment.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+    };
+}
+
+namespace aux {
```
We use `detail` or anonymous namespace for details.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+
+    dbg("worker_metrics_t::operator+() summing to proxy");
+
+    boost::for_each(worker_metrics.common_counters, [&](const counters_record_type &worker_record) {
```
Use range-based for.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+}
+
+auto
+worker_metrics_t::operator=(metrics_aggregate_proxy_t&& proxy) -> worker_metrics_t& {
```
Unobvious assignment operator usage with non-equal `this` and `other` types. Use method instead.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+                    continue;
+                }
+
+                if (uuid.compare("error") == 0) {
```
Just ==.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+
+                auto stat_it = stats_table.find(uuid);
+                if (stat_it == std::end(stats_table)) {
+                    // If isolate daemon sends us uuid we don't know about,
```
Then this is very strange. Ideally we should track uuids for slaves that were spawned somewhere here, and warn if the isolate daemon doesn't sent metrics for it for a long time.

karitra replied:
Well, _update stamp_ checking will be implemented either on this or on next iteration, stay tuned.

karitra replied:
_update stamp checking_ -  done in ebb6e99

Ready to merge in order avoid blocking engine_t (and node service in general) development, @3Hren, @antmat  PTAL. 

Stalled **TODO items** will be implemented within separate PR on _stout_ willingness.

3Hren replied:
Ok, waiting for @antmat review.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+                    // would be removed from request list and from stats table
+                    // on next poll iteration preparation.
+                    dbg("[response] inserting new metrics record with uuid" << uuid);
+                    tie(stat_it, std::ignore) = stats_table.emplace(uuid, worker_metrics_t{ctx, app_name, uuid});
```
std::tie?

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+
+    auto
+    has_errors() -> bool {
+        return ! errors.empty();
```
Space.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+    }
+
+    auto
+    errors_ref() -> const std::vector<error_t>& {
```
Just `errors`, rename field if necessary.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+private:
+    std::string app_name;
+
+    bool processed;
```
Why do we need such field?

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+            dbg("[response] metrics name: " << name);
+
+            // cut off type from name
+            const auto pos = name.find('.');
```
I feel that mangled-naming-processing stuff unnatural here. Decompose it from.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+
+                    record.value->store(incoming_value);
+                }
+            // note: e is boost::bad_get in current implementaion
```
Then catch it.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+{}
+
+metrics_retriever_t::self_metrics_t::self_metrics_t(context_t& ctx, const std::string& name) :
+   uuid_requested{ctx.metrics_hub().counter<std::uint64_t>(cocaine::format("{}.uuid_requested", name))},
```
Replace with `make_counter(ctx, name)` or something to reduce boilerplate.

3Hren commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,528 @@
+    parent->self_metrics.responses_received->fetch_add(1);
+
+    if (processor.has_errors()) {
+            const auto& errors = processor.errors_ref();
```
Something wrong with indents here.

3Hren commented on node/src/node/isometrics.hpp:
```diff
@@ -0,0 +1,274 @@
+
+        virtual
+        auto
+        despawned() -> void override {}
```
I think we should drop `spawned`/`despawned` overloads without an `uuid` parameter.

3Hren commented on node/tests/isolate_mock.cpp:
```diff
@@ -115,12 +126,81 @@ struct spawn_slot_t:
+
+        rusage ru;
+        // note: values used as dummies and don't match real metrics meanings
+        if (getrusage(RUSAGE_THREAD, &ru)) {
```
```
        #if defined(__linux__)
            getrusage(RUSAGE_THREAD, &ru);
        #else
            getrusage(RUSAGE_SELF, &ru);
        #endif
```

karitra committed:
wip: post review fixes (1)

karitra committed:
wip: worker metrics 'last updated' time stamp

Monitor worker metrics last update time stamp and signal (put to log, etc)
the cases when metrics of active worker wasn't updated for long period
of time.

antmat commented on node/src/isolate/external.cpp:
```diff
@@ -362,6 +397,22 @@ spawn_load_t::apply() {
 void
+metrics_load_t::apply() {
+    auto _inner = inner;
+    stream.apply([&](decltype(stream.unsafe())& stream){
```
why do you need synchronized here?

karitra replied:
It is taken by analogy from code of spool/spawn, mostly copy-pasted brainlessly, partly for code-base consistency. It seems that it'll not do any harm.

antmat replied:
I don't think that aquiring a mutex for code consistency is a good idea) other handles can be cancelled from other thread, and it looks like this one can not.

antmat commented on node/src/isolate/external.cpp:
```diff
@@ -362,6 +397,22 @@ spawn_load_t::apply() {
+        } catch (const std::system_error& e) {
+            COCAINE_LOG_WARNING(_inner->log, "could not process isolation metrics request: {}", error::to_string(e));
+            handle->on_error(e.code(), e.what());
+            _inner->session->detach(e.code());
```
Maybe it not really required here

karitra replied:
for code-base consistency.

antmat commented on node/src/isolate/external.cpp:
```diff
@@ -434,4 +485,25 @@ external_t::spawn(const std::string& path,
+void
+external_t::metrics(const dynamic_t& query, std::shared_ptr<api::metrics_handle_base_t> handle) const {
+
+    auto load = std::make_shared<metrics_load_t>(inner, query, handle);
```
Why do you need load here? Looks like there is no queue for metrics requests

karitra replied:
For code-base consistency with spool/spawn, initial release has metrics load queue, but it was removed as redundant for polling task, as it seems not crucial to loose some payload for it.

antmat commented on node/src/isolate/external.cpp:
```diff
@@ -434,4 +485,25 @@ external_t::spawn(const std::string& path,
+        if(inner->session && inner->prepared) {
+            COCAINE_LOG_DEBUG(inner->log, "processing metrics request");
+            load->apply();
+        } else {
```
in else block handle will never be called

antmat commented on node/src/node/engine.cpp:
```diff
@@ -86,6 +110,11 @@ engine_t::profile() const {
 }
 
+void
+engine_t::attach_pool_observer(const std::shared_ptr<pool_observer>& observer) {
```
not an issue, but why not pass by value and move?

antmat commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,594 @@
+
+    template<typename Container>
+    auto
+    dump_uuids(std::ostream& os, const std::string& logo, const Container& arr) -> void {
```
Looks like it's some kind of debug helpers. We already have newly introduced display helpers in core, you can use them, besides you can use logger to unify your debug output. If you don't want to use context_t provided logger you can do smth like https://github.com/3Hren/cocaine-plugins/blob/v0.12.13/vicodyn/include/cocaine/vicodyn/debug.hpp
@3Hren Can we think about adding such development debug functionality to core?

karitra replied:
Well, will consider it on next developing iteration.

antmat commented on node/tests/isolate_mock.cpp:
```diff
@@ -29,8 +29,27 @@
+namespace aux {
+    struct formatter_t {
+        auto
+        operator()(const std::map<std::string, std::string>& data) -> std::string {
```
Again - you can use display_traits for this

antmat commented on node/tests/isolate_mock.cpp:
```diff
@@ -115,12 +126,81 @@ struct spawn_slot_t:
+    typedef typename io::basic_slot<io::isolate::metrics>::dispatch_type dispatch_type;
+
+    virtual
+    boost::optional<std::shared_ptr<dispatch_type>>
```
result_type

antmat commented on node/tests/isolate_mock.cpp:
```diff
@@ -115,12 +126,81 @@ struct spawn_slot_t:
+    }
+
+    virtual
+    boost::optional<std::shared_ptr<dispatch_type>>
```
result_type

antmat commented on node/tests/isolate_mock.cpp:
```diff
@@ -115,12 +126,81 @@ struct spawn_slot_t:
     }
 };
 
+struct metrics_slot_t:
```
In general it's now optional to create a separate slot - you can use a function receiving headers, tuple_type, upstream_type, but it's up to you.

antmat reviewed (commented):
Still need to read and understand the main part - isometrics.[ch]pp

antmat commented on node/src/node/isometrics.hpp:
```diff
@@ -0,0 +1,278 @@
+namespace node {
+
+namespace conf {
+    constexpr auto METRICS_POLL_INTERVAL_S = 2u;
```
looks like we define constants in a similar to varible way - lower case with underscores

antmat commented on node/src/node/isometrics.hpp:
```diff
@@ -0,0 +1,278 @@
+    constexpr auto METRICS_POLL_INTERVAL_S = 2u;
+}
+
+enum class CounterType : unsigned {
```
style

antmat commented on node/src/node/isometrics.hpp:
```diff
@@ -0,0 +1,278 @@
+{
+    const auto node_config = ctx.config().component_group("services").get("node");
+
+    if (!node_config) {
```
is this a good idea to return nullptr? Besides for some(very strange) reasons node service can have a different name (not 'node').

antmat commented on node/src/node/isometrics.cpp:
```diff
@@ -0,0 +1,594 @@
+namespace ph = std::placeholders;
+
+namespace conf {
+    constexpr auto PURGATORY_QUEUE_BOUND = 8 * 1024;
```
seems that lowercase is better in this case

karitra committed:
wip: post-review fixes, part I

antmat reviewed (approved):
Looks ok, I'll try to take a deeper look later

3Hren committed:
Merge branch 'master' into workers_metrics
